### PR TITLE
APIE-1063: Migrate docker/docker to moby/moby to fix CVE-2026-34040 and CVE-2026-33997

### PR DESCRIPTION
**Pull request by linouk23**
Release Notes
-------------

Intentionally empty. This is a dependency-only refactor with no user-visible behavior change — no command, flag, output column, or serialized field is added, renamed, or removed. The release-notes history in [`confluentinc/cli-release/release-notes/`](https://github.com/confluentinc/cli-release/tree/master/release-notes) does not call out CVE fixes or dependency bumps (137 files surveyed; 0 references to `CVE`, `vulnerability`, `security fix`, or `docker/docker`), so this PR follows the same convention.

Checklist
---------

- [x] I have successfully built and used a custom CLI binary, without linter issues from this PR.
- [x] I have clearly specified in the `What` section below whether this PR applies to Confluent Cloud, Confluent Platform, or both.
- [x] I have verified this PR in Confluent Cloud pre-prod or production environment, if applicable.
- [x] I have verified this PR in Confluent Platform on-premises environment, if applicable.
- [x] I have attached manual CLI verification results or screenshots in the `Test & Review` section below.
- [x] I have added appropriate CLI integration or unit tests for any new or updated commands and functionality.
- [x] I confirm that this PR introduces no breaking changes or backward compatibility issues.
- [x] I have indicated the potential customer impact if something goes wrong in the `Blast Radius` section below.
- [x] I have put checkmarks below confirming that the feature associated with this PR is enabled in:
  - [x] Confluent Cloud prod
  - [ ] Confluent Cloud stag
  - [x] Confluent Platform
  - [ ] Check this box if the feature is enabled for certain organizations only

What
----
**Scope.** The touched code (`internal/local/command_kafka{,_start,_stop}.go`) powers `confluent local kafka start/stop` — a **laptop-only dev sandbox** that spins up Confluent's experimental [`confluentinc/confluent-local:8.0.0`](https://hub.docker.com/r/confluentinc/confluent-local) image (KRaft mode, "[built for local development workflows and is not officially supported for production workloads](https://hub.docker.com/r/confluentinc/confluent-local)") via the user's local Docker daemon. The CLI docs frame it as a "[single-node development environment](https://docs.confluent.io/confluent-cli/current/command-reference/local/kafka/confluent_local_kafka_start.html)" — it does **not** call the Confluent Cloud control plane and is **not** part of a Confluent Platform install.

So "applies to both Cloud and Platform" is trivially true — the same binary ships to both audiences and either can run `local kafka` on their laptop — but doesn't mean there are two surfaces to test. `confluent local kafka` is a third bucket in this repo: alongside [`RequireCloudLogin`](pkg/cmd/ANNOTATIONS.md)-gated commands and `RequireOnPremLogin`-gated commands, it carries **no `RunRequirement` annotation at all** (verified: `grep -rn Annotations internal/local/` → zero hits) and works without any login. Reviewers can validate this PR with **one Docker daemon, in one shell, in one pass** — there is no CC-vs-CP code split to exercise.

**Problem.** `CVE-2026-34040` (AuthZ plugin bypass with oversized request body) and `CVE-2026-33997` (off-by-one in `docker plugin install` privilege validation) are patched only in Docker Engine **29.3.1**. moby v29 renamed the module root to `github.com/moby/moby/v2` and split the SDK into separate `github.com/moby/moby/api` and `github.com/moby/moby/client` modules. **Nothing is published as `github.com/docker/docker@v29.x`** (`go get github.com/docker/docker@v29.3.1+incompatible` → `not found: invalid version`), and v28 has no backport. So this is a **module migration**, not a version bump — same shape as the go-jose v3→v4 work in #3363.

**Exploitability of these CVEs in the CLI.** Both are **daemon-side** code paths (the AuthZ middleware and the `docker plugin install` privilege checker). The CLI is a pure SDK consumer talking to a remote `dockerd`; the vulnerable code never executes inside the CLI process. This PR is a scanner-satisfaction fix, not a runtime-risk fix.

**Sources of `docker/docker` in this repo (pre-PR).**
| Importer | Where | How it's used |
|---|---|---|
| Direct | `internal/local/command_kafka.go` | `*client.Client`, `Info` |
| Direct | `internal/local/command_kafka_start.go` | `client.NewClientWithOpts`, `ContainerList`, `ImagePull`, `NetworkCreate`, `ContainerCreate`, `ContainerStart`, plus `api/types/{container,image,network,strslice}` |
| Direct | `internal/local/command_kafka_stop.go` | `ContainerList`, `ContainerStop`, `ContainerRemove` |
| Transitive (test-only) | `confluent-kafka-go/v2/kafka` test files → `testcontainers-go/modules/compose` | not compiled into our binary — see References (https://github.com/confluentinc/confluent-kafka-go/pull/1554, https://github.com/confluentinc/confluent-kafka-go/issues/1547) |

**Implementation.**
1. Swap 5 imports across the three `command_kafka*.go` files: `github.com/docker/docker/{client,api/types/{container,image,network,strslice}}` → `github.com/moby/moby/{client,api/types/{container,network,strslice}}`. The `image` import disappears because `ImagePullOptions` moved into the `client` package.
2. v29 reorganized the client API. Refactor the ~10 affected call sites:
   - Per-call options moved into `client`: `client.ContainerListOptions`, `client.ImagePullOptions`, `client.NetworkCreateOptions`, `client.ContainerStartOptions`, `client.ContainerStopOptions`, `client.ContainerRemoveOptions`, `client.InfoOptions`.
   - `ContainerList` now returns `ContainerListResult` — iterate `.Items`.
   - `ContainerCreate` now takes a single `ContainerCreateOptions` struct (`Config`/`HostConfig`/`Platform`/`Name`) instead of 5 positional args; `NetworkingConfig` defaults to nil by omission.
   - `Container{Start,Stop,Remove}` and `Info` now return a `Result` struct alongside `error` — discard with `_,`.
3. Drop `github.com/docker/go-connections/nat`. Port types moved to `github.com/moby/moby/api/types/network` (`PortSet`, `PortMap`, `PortBinding`, `Port`). `network.PortBinding.HostIP` is now `netip.Addr` not `string`, so wrap the existing `localhost = "0.0.0.0"` constant at the two call sites with `netip.MustParseAddr` (`netip.Addr` JSON-marshals to `"0.0.0.0"`, so the wire format is preserved). `nat.Port(s)` (a string type conversion) becomes `network.MustParsePort(s)`; inputs are already validated upstream by `net.Listen` + `freeport.GetFreePort` so panic is precluded in practice.
4. `go mod tidy` drops `github.com/docker/docker` from `require`, adds `github.com/moby/moby/api v1.54.2` and `github.com/moby/moby/client v0.4.1`. Routine transitive bumps follow: `containerd/errdefs v0.1.0 → v1.0.0` (indirect), `docker/go-connections v0.5.0 → v0.7.0`, `opencontainers/image-spec v1.1.0 → v1.1.1`, etc.

**Why not a `replace github.com/docker/docker => github.com/moby/moby/v2 ...` shortcut?** Inside moby/moby's v29 repo, `api/` and `client/` each have their own `go.mod` — they're **nested modules**, not subdirectories of `github.com/moby/moby/v2`. A `replace` of the root wouldn't reach those import subpaths, so we'd need separate shims for both `/api` and `/client`. Direct source migration is cleaner for our usage (the splunk PR linked in References had to use a shim only because their direct callers were transitive deps).

Blast Radius
----

`internal/local/command_kafka*.go` is the **only consumer of docker/docker in our module graph**. The only customer-facing surface affected is the `confluent local kafka` family:

- **If a v29-client request is wire-encoded differently than v28** — e.g., `PortBinding.HostIp` not serializing to the expected string, `ContainerCreate`'s new options-struct mapping to a different request body — then `confluent local kafka start` will fail to create the broker container (no Kafka REST port becomes available) and `confluent local kafka stop` will fail to enumerate/stop existing containers. Customers using `confluent local kafka` to spin up a dev cluster on their laptop would see container-create errors at start time.
- The structural risk is small: `network.PortBinding.HostIP` is `netip.Addr` with `json:"HostIp"` tag and `MarshalText() → "0.0.0.0"`; `network.PortSet`/`PortMap` are type aliases for the same underlying maps; the moby/api package is specifically designed for wire compatibility with dockerd. But the only way to be 100% sure is a live smoke test — see Test & Review.
- No other CLI command path touches docker/docker, so all non-`local` commands (`confluent kafka …`, `confluent api-key …`, `confluent connect …`, etc.) are unaffected. No CI surface, no installer, no packaging is touched.

No customer surface changes (no flag, no output column, no serialized field added/renamed/removed).

References
----------

- Jira: [APIE-1063](https://confluentinc.atlassian.net/browse/APIE-1063)
- CVEs:
  - [CVE-2026-34040](https://www.cve.org/CVERecord?id=CVE-2026-34040) — AuthZ plugin bypass · GHSA: [x744-4wpc-v9h2](https://github.com/moby/moby/security/advisories/GHSA-x744-4wpc-v9h2)
  - [CVE-2026-33997](https://www.cve.org/CVERecord?id=CVE-2026-33997) — plugin install privilege validation off-by-one · GHSA: [pxq6-2prw-chj9](https://github.com/moby/moby/security/advisories/GHSA-pxq6-2prw-chj9)
- moby v29 restructuring: [docker-v29.3.1 release notes](https://github.com/moby/moby/releases/tag/docker-v29.3.1), [`go.mod` at v29.3.1 declaring `module github.com/moby/moby/v2`](https://github.com/moby/moby/blob/docker-v29.3.1/go.mod), [`moby/moby/api`](https://pkg.go.dev/github.com/moby/moby/api) + [`moby/moby/client`](https://pkg.go.dev/github.com/moby/moby/client) module docs
- Related Confluent PRs:
  - [confluentinc/cli#3363](https://github.com/confluentinc/cli/pull/3363) — APIE-1061 go-jose v3→v4, same module-migration shape
  - **[confluentinc/confluent-kafka-go#1535](https://github.com/confluentinc/confluent-kafka-go/pull/1535) — the upstream path to dropping our last `go.sum` reference to docker/docker.** It bumps `testcontainers-go/modules/compose` from v0.33.0 to v0.42.0 (which migrated to moby for the main package), but `testcontainers-go/modules/compose` itself still pulls `docker/docker` — the PR is **currently OPEN with merge conflicts**, stalled on whether to drop the `compose` submodule entirely (would require running `docker compose` via subprocess). Tracking issue: [confluent-kafka-go#1547](https://github.com/confluentinc/confluent-kafka-go/issues/1547). When that PR merges and a release is cut, a routine `go mod tidy` on this repo will remove the lingering go.sum line.
- Prior art for the same docker→moby migration in other ecosystems:
  - [signalfx/splunk-otel-collector#7521](https://github.com/signalfx/splunk-otel-collector/pull/7521) — closest analog (same CVE-driven motivation; uses `replace`+shim because their direct callers were transitive — see "Why not `replace`" above)
  - [elastic/elastic-agent#13804](https://github.com/elastic/elastic-agent/pull/13804), [DataDog/datadog-agent#48777](https://github.com/DataDog/datadog-agent/pull/48777), [grafana/alloy#6167](https://github.com/grafana/alloy/pull/6167) — direct source-migration approaches

Test & Review
-------------
Manual verification: [link](https://confluentinc.atlassian.net/wiki/spaces/AEGI/pages/5627216915/1-Pager+Manual+Verification+of+PR+3364+docker+docker+moby+moby+migration)

- `make lint-go` — clean (`✅ golangci-lint`)
- `go build ./...` — clean
- `go vet ./internal/local/...` — clean
- `go test ./internal/local/... ./internal/asyncapi/... -count=1` — pass (asyncapi is tested because it's the only other package still reaching the lingering docker/docker test-dep via `confluent-kafka-go`'s test files)
- Smoke: built `./cmd/confluent`, `--version` and `local kafka --help` produce expected command tree
- **Binary symbol check** confirming docker/docker is not linked: `go tool nm /tmp/confluent-built | grep -E 'github.com/docker/docker(/|$)'` → **zero** matches; `go tool nm | grep -c 'github.com/moby/moby'` → **366** symbols present
- `grep 'github.com/docker/docker' go.mod` → empty (the `require` block is fully clean; the only lingering reference is 2 lines in `go.sum` for a test-only transitive — see References)
- `network.ParsePort` edge-case probe — ports `"0/tcp"`, `"8082/tcp"` parse cleanly; `"65536/tcp"`, `"abc/tcp"`, `""` return errors. Our `MustParsePort` call sites only fire after `net.Listen` / `freeport.GetFreePort` validation, so panic is precluded.
- **Manual `confluent local kafka` smoke against a real Docker daemon: PASSED.** Six test cases executed end-to-end against the locally-built PR binary and a running Docker daemon — TC-01 command-tree smoke, TC-02 single-broker happy path with REST `curl`, TC-03 multi-broker × 3, TC-04 custom-port `HostIp` wire-format check, TC-05 idempotent double-start / double-stop, TC-06 daemon-down negative path; all passed. Detailed commands and observed output captured on the [manual-verification page](https://confluentinc.atlassian.net/wiki/spaces/AEGI/pages/5627216915/1-Pager+Manual+Verification+of+PR+3364+docker+docker+moby+moby+migration) linked at the top of this section. v29 wire-format compatibility for `PortBinding`/`ContainerCreate`/`ImagePull` is confirmed against a live daemon: `0.0.0.0` port binding round-trips correctly (`docker port` shows `0.0.0.0:<host-port>` for both REST and plaintext bindings) and the full container lifecycle (create → start → list → stop → remove) works under the new moby client.

[APIE-1063]: https://confluentinc.atlassian.net/browse/APIE-1063

### CODE REVIEW

File: internal/local/command_kafka.go [hunk 1]
```diff
@@ -3,7 +3,7 @@ package local
 import (
 	"context"
 
-	"github.com/docker/docker/client"
+	"github.com/moby/moby/client"
 	"github.com/spf13/cobra"
```
[the remark this answers is not in the document]
Comment (linouk23) [reply internal/local/command_kafka.go:7]: The review fired at 08:51 against the initial placeholder state. Title and body were rewritten shortly after — concerns are now addressed in the current description:

- **Why migrate** (vs. a `go get` bump): **What → Problem.** moby v29 renamed the module root to `github.com/moby/moby/v2` and split the SDK into `moby/moby/api` + `moby/moby/client`; `github.com/docker/docker@v29.x` doesn't exist as a Go module, and v28 has no backport for either CVE.
- **Customer impact for `confluent local kafka`**: **Blast Radius.** Scope is exactly `confluent local kafka start/stop` (the only consumers of docker/docker in our module graph); no other CLI surface is touched. Likely failure mode and structural reasons it shouldn't trigger are spelled out.
- **Release Notes**: intentionally empty with rationale, matching the cli-release/release-notes convention (137 files surveyed, 0 references to CVE/dep bumps).
- **Exploitability**: both CVEs are daemon-side; the CLI is a pure SDK consumer, so the vulnerable code never runs in our process — this PR is a scanner-satisfaction fix, not a runtime-risk fix. Called out in **What → Exploitability**.

**Comment by stefanb** [comment]
Similar will have to be done in https://github.com/confluentinc/confluent-kafka-go because old module is being referenced there:
https://github.com/confluentinc/confluent-kafka-go/blob/561ea1a6f5e0d0a8d36e9fbfbfb444f63d5536c1/go.mod#L86

via https://github.com/testcontainers/testcontainers-go
https://github.com/confluentinc/confluent-kafka-go/blob/561ea1a6f5e0d0a8d36e9fbfbfb444f63d5536c1/kafka/integration_test.go#L33

which then propagated to downstream Go modules because integration tests are in the same package
* https://github.com/confluentinc/confluent-kafka-go/issues/1547

**Comment by linouk23** [comment]
### Re: SonarQube "0.0% Coverage on New Code" — requesting Quality Gate override

The 0% number is misleading for this PR's diff shape. The diff **does** include a new unit test — it just only covers the one pure-Go helper in the change; the rest of the diff is structurally daemon-dependent and has never been unit-testable in this repo.

**What this PR adds in tests.** `TestGetNatPlaintextPorts` (commit [`a52f2f71`](https://github.com/confluentinc/cli/pull/3364/commits/a52f2f71), `internal/local/command_kafka_start_test.go`) — table-driven, three cases (empty, single broker, multi-broker), exercising the new moby `network.Port` / `network.MustParsePort` round-trip. Takes `getNatPlaintextPorts` from 0% → **100%** coverage. It's the only function in the diff that doesn't require a running Docker daemon.

**Why the percentage is still low despite the new test.** Sonar's "% on new code" weighs every changed line equally. ~22–25 of the changed lines live inside functions that call `dockerClient.{Info, ContainerList, ImagePull, NetworkCreate, ContainerCreate, ContainerStart, ContainerStop, ContainerRemove}`. Those lines aren't unit-testable without either (a) introducing a `dockerClient` interface + hand-rolled mocks across the production surface, or (b) requiring a real Docker daemon in CI. Both are scope expansion beyond a CVE migration.

**Baseline.** Per-function picture on `main` *before* this PR vs. after — every function in the touched files was 0% on `main`:

| Function | On `main` | After this PR |
|---|---|---|
| `kafkaStart` | 0.0% | 0.0% |
| `kafkaStop` | 0.0% | 0.0% |
| `checkIsDockerRunning` | 0.0% | 0.0% |
| `stopAndRemoveConfluentLocal` | 0.0% | 0.0% |
| `getNatPlaintextPorts` | 0.0% | **100%** ← new test |
| `getContainerEnvironmentWithPorts` | 0.0% | 0.0% |
| `getKafkaControllerQuorumVoters` | 0.0% | 0.0% |
| `getKafkaRestBootstrapServers` | 0.0% | 0.0% |
| `checkMachineArch` | 0.0% | 0.0% |

Package-level coverage was 10.6%, all of which comes from `command_service*_test.go` covering the pre-Docker `confluent local services` path. There is no Docker client mock or `httptest` daemon fake anywhere in this repo — never has been.

**Manual verification (in lieu of integration tests).** Six end-to-end test cases executed against a live Docker daemon and the locally-built PR binary — TC-01 command-tree smoke, TC-02 single-broker happy path with REST `curl`, TC-03 multi-broker × 3, TC-04 custom-port `HostIp` wire-format check, TC-05 idempotent double-start / double-stop, TC-06 daemon-down negative; all passed. v29 wire-format compatibility for `PortBinding`/`ContainerCreate`/`ImagePull` confirmed against a live daemon. Detail and observed output on the [manual-verification page](https://confluentinc.atlassian.net/wiki/spaces/AEGI/pages/5627216915/1-Pager+Manual+Verification+of+PR+3364+docker+docker+moby+moby+migration) (also linked from the Test & Review section of the PR body).

**Precedent.** Prior `docker/docker` bumps on these same files — [#2814](https://github.com/confluentinc/cli/pull/2814) 26.1.4→27, [#2726](https://github.com/confluentinc/cli/pull/2726) 25.0.5→26, [#2654](https://github.com/confluentinc/cli/pull/2654) 24.0.7→25 — all shipped without coverage when Sonar gating wasn't yet enabled on this repo. The recent [#3363](https://github.com/confluentinc/cli/pull/3363) (go-jose v3→v4 — same CVE-migration pattern) passed Sonar only because `pkg/config` already had tests touching the modified function. This PR doesn't have that lucky property through no fault of the diff.

**Request.** Override the Quality Gate. The diff is a security-driven CVE fix ([CVE-2026-34040](https://www.cve.org/CVERecord?id=CVE-2026-34040), [CVE-2026-33997](https://www.cve.org/CVERecord?id=CVE-2026-33997)); the one pure-Go helper in the diff *is* covered by a new unit test; the residual uncovered lines are structurally daemon-dependent; manual verification against a live Docker daemon has passed; and the entire `confluent local kafka` test surface has always been integration-test territory.

If override is denied, happy to follow up with either:
- A smaller refactor extracting pure-logic helpers (`buildBrokerConfig`, `buildBrokerHostConfig`, `buildContainerCreateOptions`) from the `kafkaStart` body — gets to ~50% on changed lines, no behavior change.
- A full `dockerClient` interface + mocks — gets to ~80% but threads a new dependency through the production surface.

**Review by sgagniere — APPROVED** [review]
